tests/activity_api_reset_test.go

```diff
@@ -0,0 +1,325 @@
+
+	s.activityRetryPolicy = &temporal.RetryPolicy{
+		InitialInterval:    s.initialRetryInterval,
+		BackoffCoefficient: 1,
+	}
+}
+
+func TestActivityApiResetClientTestSuite(t *testing.T) {
+	s := new(ActivityApiResetClientTestSuite)
+	suite.Run(t, s)
+}
+
+func (s *ActivityApiResetClientTestSuite) makeWorkflowFunc(activityFunction ActivityFunctions) WorkflowFunction {
+	return func(ctx workflow.Context) (string, error) {
+
+		var ret string
+		err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			ActivityID:             "activity-id",
+			DisableEagerExecution:  true,
+			StartToCloseTimeout:    s.startToCloseTimeout,
+			ScheduleToCloseTimeout: s.scheduleToCloseTimeout,
+			RetryPolicy:            s.activityRetryPolicy,
+		}), activityFunction).Get(ctx, &ret)
+		return "done!", err
+	}
+}
+
+func (s *ActivityApiResetClientTestSuite) TestActivityResetApi_AfterRetry() {
+	// activity reset is called after multiple attempts,
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	var activityWasReset atomic.Bool
+	activityCompleteCh := make(chan struct{})
+	var activityCompleteCount atomic.Int32
+
+	activityFunction := func() (string, error) {
+		activityCompleteCount.Add(1)
+
+		if activityWasReset.Load() == false {
+			activityErr := errors.New("bad-luck-please-retry")
+			return "", activityErr
+		}
+
+		s.WaitForChannel(ctx, activityCompleteCh)
+		return "done!", nil
+	}
+
+	workflowFn := s.makeWorkflowFunc(activityFunction)
+
+	s.Worker().RegisterWorkflow(workflowFn)
+	s.Worker().RegisterActivity(activityFunction)
+
+	wfId := testcore.RandomizeStr("wfid-" + s.T().Name())
+	workflowOptions := sdkclient.StartWorkflowOptions{
+		ID:        wfId,
+		TaskQueue: s.TaskQueue(),
+	}
+
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	// wait for activity to start/fail few times
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.True(t, activityCompleteCount.Load() > 1)
+	}, 5*time.Second, 200*time.Millisecond)
+
+	resetRequest := &workflowservicepb.ResetActivityByIdRequest{
+		Namespace:  s.Namespace(),
+		WorkflowId: workflowRun.GetID(),
+		ActivityId: "activity-id",
+		NoWait:     true,
+	}
+	resp, err := s.FrontendClient().ResetActivityById(ctx, resetRequest)
+	s.NoError(err)
+	s.NotNil(resp)
+
+	activityWasReset.Store(true)
+
+	// wait for activity to be running
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_STARTED, description.PendingActivities[0].State)
```

Comment:
You can get rid of the Eventually block and just check that the attempt was reset. That should be immediate.

The state transitioning to STARTED can be eventual and has no value for this test. You already check that the activity completes so it must transition to STARTED to do that.

Reply:
no I can't get rid of this Eventually block. 
To check that the number of attempts was reset I need to call DescribeWorkflowExecution.
And this may not happen instantly after ResetActivityById call